src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -202,9 +284,173 @@ public Task LoadAsync()
 
         public Task UnloadAsync()
         {
-            _enabled = false;
+            lock (SyncObject)
+            {
+                _enabled = false;
+
+                _whenNominatedTask?.TrySetCanceled();
+            }
 
             return Task.CompletedTask;
         }
+
+        // NuGet calls this method to wait project to nominate restoring.
+        // If the project has no pending restore data, it will return a completed task.
+        // Otherwise a task which will be completed once the project nominate the next restore
+        // the task will be cancelled, if the project system decide it no longer need restore (for example: the restore state has no change)
+        // the task will be failed, if the project system runs into a problem, so it cannot get correct data to nominate a restore (DT build failed)
+        public Task WhenNominated(CancellationToken cancellationToken)
+        {
+            lock (SyncObject)
+            {
+                if (!CheckIfHasPendingNomination() || cancellationToken.IsCancellationRequested)
+                {
+                    return Task.CompletedTask;
+                }
+
+                _whenNominatedTask ??= new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
I think this logic would be simpler if; _whenNominatedTask just always had a value and we never had to worry about null paths, or handling that we haven't registered the ContinueWith. It simplifies the amount of state and the possible values of those states. @lifengl

Reply:
At the beginning, I thought that too, and suggested to use AsyncAutoResetEvent, but moved away from it later. There are couple reasons. One thing initially is that nuget originally said they would not use WhenNorminate in the initial implementation, so creating them would be a waste. In the end, they do use it, but it would just block on first project returns an unfinished task, we don't need create the structure for most other projects.  And later, more importantly, it might make it more confusing, due to couple design issues and special race condition handling, which would make the always existing task completion source to have a different state than the WhenNominated state, which can lead confusion especially during dump analysis.

Two of most important problems related to that:
1, the component actually doesn't know the start time WhenNominated returns an unfinished task. It doesn't actively monitor project changes, but only passively checks it when it is asked for. (It actually reduced overhead in runtime). So other design might lead to create incomplete task source after one nomination is done, even we may never have one in the future?

2, the project can be changed at the same time this component nominates restore. we don't want the same project to block its own restore request, or it will lead a deadlock (maybe we should cancel it, but it is more complex, and I intend to leave it to the future). To prevent that, nominating restore always means that a, HasPendingNomination will always returns false, until the request is done, b, it always finishes any previous waiting task, even the project version is changed again to make the request out-of-dated. This race condition is the primary reason to lead some complication in the implementation.